### PR TITLE
[to delete] create family model

### PR DESCRIPTION
**Pull request by scara82**
To be replaced with ms-add-family-model (https://github.com/hooroo/domus/pull/7)
 Family model and simple form

### CODE REVIEW

File: app/controllers/families_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  def new
+    @family = Family.new
+  end
+
+  def create
+    family = Family.create(family_params)
+    if family.save
+      redirect_to root_path
+    else
+      redirect_to "/"
```
Comment (elle) [inline app/controllers/families_controller.rb]: We usually do `render :new` here, where we display validation errors.
We also usually set a flash message here when something goes wrong, so the user is not left wondering as to what happened.

File: app/controllers/families_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    end
+  end
+
+  def destroy
+  end
```
Comment (elle) [inline app/controllers/families_controller.rb]: Since `destroy` is not implemented here, I would remove it from this PR

File: app/models/family.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class Family < ApplicationRecord
+  validates_presence_of :last_name, :capability
+  #polimorphism implementation
+  # has_many: phone_numbers, as: :contact
```
Comment (elle) [inline app/models/family.rb:4]: Remove comments??

File: app/views/families/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title></title>
+  </head>
+  <body>
```
Comment (elle) [inline app/views/families/index.html.erb]: You don't need to set all those tags, because they are defined in `app/views/layouts/application.html.erb`
If you look at the families index view in the browser, I believe you will see those tags defined twice.

File: app/views/families/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+<!DOCTYPE html>
+<html>
+  <head>
+    <meta charset="utf-8">
+    <title></title>
+  </head>
+  <body>
```
Comment (elle) [inline app/views/families/new.html.erb]: Same as above

File: app/views/families/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    <%= simple_form_for @family do |f| %>
+      <%= f.input :first_name, placeholder: 'first_name' %>
+      <%= f.input :last_name, placeholder: 'last_name' %>
+      <%= f.input :address, placeholder: 'address' %>
+      <%= f.input :animals, placeholder: 'animals' %>
+      <%= f.input :capability, placeholder: 'capability' %>
+      <%= f.input :prof, as: :radio_buttons, collection: ['Y', 'N'] %>
+      <%= f.input :preferred_gender, as: :radio_buttons, collection: ['F', 'M'] %>
+      <%= f.input :bus_driver, as: :radio_buttons, collection: ['Y', 'N'] %>
+      <%= f.input :active, as: :radio_buttons, collection: ['Y', 'N'] %>
```
Comment (elle) [inline app/views/families/new.html.erb]: Double quotes

File: config/application.rb [hunk 7]
```diff
@@ -15,6 +15,11 @@
 
 module Domus
   class Application < Rails::Application
+
+    config.middleware.use ::Rack::Auth::Basic do |u, p|
+      [u, p] == [ENV["BASIC_AUTH_USERNAME"], ENV["BASIC_AUTH_PASSWORD"]]
+    end
+
```
Comment (elle) [inline config/application.rb:25]: This does not belong in this PR

File: config/locales/en.yml [hunk 8]
```diff
@@ -1,33 +1,2 @@
...
-# true, false, on, off, yes, no
-#
-# Instead, surround them with single quotes.
-#
-# en:
-#   'true': 'foo'
-#
-# To learn more, please read the Rails Internationalization guide
-# available at http://guides.rubyonrails.org/i18n.html.
-
```
Comment (elle) [inline config/locales/en.yml:31]: This change is from a different PR

File: config/locales/simple_form.en.yml [hunk 9]
```diff
@@ -0,0 +1,31 @@
+en:
+  simple_form:
+    "yes": 'Yes'
+    "no": 'No'
+    required:
+      text: 'required'
+      mark: '*'
```
Comment (elle) [inline config/locales/simple_form.en.yml]: Either use double quotes or no quotes

File: config/routes.rb [hunk 10]
```diff
@@ -1,3 +1,5 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  resources :families
```
Comment (elle) [inline config/routes.rb]: Since you only implemented index, new, and create, I would restrict the routes here to be:
`resources :families, only: [:index, :new, :create]`

File: db/migrate/20171208043003_create_families.rb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    create_table :families do |t|
+      t.string :first_name
+      t.string :last_name, null: false
+      t.string :address
+      t.string :animals
+      t.integer :capability, null: false
+      t.boolean :prof
+      t.string :preffered_gender
+      t.boolean :bus_driver
+      t.boolean :active
```
Comment (elle) [inline db/migrate/20171208043003_create_families.rb:12]: We probably want to set the default here to be true

File: db/migrate/20171208043003_create_families.rb [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  def change
+    create_table :families do |t|
+      t.string :first_name
+      t.string :last_name, null: false
+      t.string :address
+      t.string :animals
+      t.integer :capability, null: false
+      t.boolean :prof
+      t.string :preffered_gender
+      t.boolean :bus_driver
```
Comment (elle) [inline db/migrate/20171208043003_create_families.rb:11]: We probably want to set the default here to false

File: db/migrate/20171208043003_create_families.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+class CreateFamilies < ActiveRecord::Migration[5.1]
+  def change
+    create_table :families do |t|
+      t.string :first_name
+      t.string :last_name, null: false
+      t.string :address
+      t.string :animals
+      t.integer :capability, null: false
+      t.boolean :prof
```
Comment (elle) [inline db/migrate/20171208043003_create_families.rb:9]: We probably want to set the default here to false

File: spec/controllers/families_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (elle) [inline spec/controllers/families_controller_spec.rb]: Double quotes

File: spec/models/family_spec.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
```
Comment (elle) [inline spec/models/family_spec.rb]: Double quotes

File: spec/models/family_spec.rb [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+describe Family do
+  describe "#create" do
+    it "is not valid without a last_name" do
+      new_family = Family.new(last_name: nil)
+
+      expect(new_family).to_not be_valid
```
Comment (elle) [inline spec/models/family_spec.rb]: We prefer to use `.not_to` because it reads better

File: spec/models/family_spec.rb [hunk 17]
```diff
@@ -0,0 +1,27 @@
+require 'rails_helper'
+
+describe Family do
+  describe "#create" do
+    it "is not valid without a last_name" do
+      new_family = Family.new(last_name: nil)
```
Comment (elle) [inline spec/models/family_spec.rb]: I like that you are explicit here, but just calling `.new` would have suffice.

Also, look at using shoulda matchers for all the validations. Make life easier.

File: spec/models/family_spec.rb [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+      expect(new_family).to_not be_valid
+    end
+
+    it "is not valid without capability" do
+      new_family = Family.new(capability: nil)
+
+      expect(new_family).to_not be_valid
```
Comment (elle) [inline spec/models/family_spec.rb]: Same as above

**Review by elle — COMMENTED** [review]
PR description could use some help.

I would use shoulda-matchers for validations and put all the validations in a describe block, i.e.:
```rb
describe "validations" do
end
```

I would also add a family factory to `spec/factories.rb` with the minimum required attributes to make the factory valid.

There are files in this PR that belong to a different PR and not this one. I would be more picky with the changes I include in a PR.

And lastly, feature specs are missing for the functionality to see all families, creating a new family.
I would also add a controller spec when trying to create an invalid family, to see the flash message that was set, an errors on the family object

**Comment by elle** [comment]
A suggestion to look at the responders gem, and see if you wish to use it or not

Ref: https://github.com/plataformatec/responders

File: spec/rails_helper.rb [hunk 19]
```diff
@@ -1,14 +1,20 @@
...
 Dir[Rails.root.join("spec/support/**/*.rb")].each { |f| require f }
 
 ActiveRecord::Migration.maintain_test_schema!
 
+Shoulda::Matchers.configure do |config|
+  config.integrate do |with|
+    with.test_framework :rspec
+    with.library :rails
+  end
+end
```
Comment (elle) [inline spec/rails_helper.rb]: We don't add new things to this file. We add new settings in `spec/support/*`

Also, you have shoulda-matchers set up already at: https://github.com/hooroo/domus/blob/master/spec/support/shoulda_matchers.rb

File: app/views/families/index.html.erb (hunk 4 above)
Comment (scara82) [reply app/views/families/index.html.erb]: 👍

File: app/views/families/new.html.erb (hunk 5 above)
Comment (scara82) [reply app/views/families/new.html.erb]: 👍

File: app/views/families/new.html.erb (hunk 6 above)
Comment (scara82) [reply app/views/families/new.html.erb]: 👍

File: config/locales/simple_form.en.yml (hunk 9 above)
Comment (scara82) [reply config/locales/simple_form.en.yml]: 👍

File: spec/controllers/families_controller_spec.rb (hunk 14 above)
Comment (scara82) [reply spec/controllers/families_controller_spec.rb]: 👍

File: spec/models/family_spec.rb (hunk 15 above)
Comment (scara82) [reply spec/models/family_spec.rb]: 👍

File: spec/rails_helper.rb (hunk 19 above)
Comment (scara82) [reply spec/rails_helper.rb]: 👍

File: config/routes.rb (hunk 10 above)
Comment (scara82) [reply config/routes.rb]: 👍

File: db/migrate/20171208043003_create_families.rb (hunk 13 above)
Comment (scara82) [reply db/migrate/20171208043003_create_families.rb:9]: 👍

File: db/migrate/20171208043003_create_families.rb (hunk 12 above)
Comment (scara82) [reply db/migrate/20171208043003_create_families.rb:11]: 👍

File: db/migrate/20171208043003_create_families.rb (hunk 11 above)
Comment (scara82) [reply db/migrate/20171208043003_create_families.rb:12]: 👍

File: app/controllers/families_controller.rb (hunk 1 above)
Comment (scara82) [reply app/controllers/families_controller.rb]: 👍
